sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala

```diff
@@ -103,6 +116,123 @@ class FileScanRDD(
+            val unsafeRowConverter = {
+              val converter = UnsafeProjection.create(
+                Array(FILE_METADATA_COLUMNS.dataType))
+              (row: Row) => {
+                converter(CatalystTypeConverters.convertToCatalyst(row)
+                  .asInstanceOf[InternalRow])
+              }
+            }
+            metadataStructUnsafeRow =
+              unsafeRowConverter(Row.fromSeq(Seq(metadataStructGenericRow)))
+          }
+        }
+
+      /**
+       * Create a writable column vector containing all required metadata fields
+       */
+      private def createMetadataStructColumnVector(
+          c: ColumnarBatch, meta: AttributeReference): WritableColumnVector = {
+        val columnVector = createColumnVector(c.numRows(), FILE_METADATA_COLUMNS.dataType)
+        val filePathBytes = new File(currentFile.filePath).toString.getBytes
+        val fileNameBytes = currentFile.filePath.split("/").last.getBytes
+        var rowId = 0
+
+        assert(meta.dataType.isInstanceOf[StructType])
+        meta.dataType.asInstanceOf[StructType].names.zipWithIndex.foreach { case (name, ind) =>
+          name match {
+            case FILE_PATH =>
+              rowId = 0
+              // use a tight-loop for better performance
+              while (rowId < c.numRows()) {
+                columnVector.getChild(ind).putByteArray(rowId, filePathBytes)
+                rowId += 1
+              }
+            case FILE_NAME =>
+              rowId = 0
+              // use a tight-loop for better performance
+              while (rowId < c.numRows()) {
+                columnVector.getChild(ind).putByteArray(rowId, fileNameBytes)
+                rowId += 1
+              }
+            case FILE_SIZE =>
+              columnVector.getChild(ind).putLongs(0, c.numRows(), currentFile.fileSize)
+            case FILE_MODIFICATION_TIME =>
+              columnVector.getChild(ind).putLongs(0, c.numRows(), currentFile.modificationTime)
+            case _ => // be exhaustive, won't happen
+          }
+        }
+        columnVector
+      }
+
+      /**
+       * Add metadata struct at the end of nextElement if needed.
+       * For different row implementations, use different methods to update and append.
+       */
+      private def addMetadataStructIfNeeded(nextElement: Object): Object =
+        if (metadataStruct.exists(_.sameRef(FILE_METADATA_COLUMNS))) {
+          val meta = metadataStruct.get
+          nextElement match {
+            case c: ColumnarBatch =>
+              val columnVectorArr = Array.tabulate(c.numCols())(c.column) ++
+                Array(createMetadataStructColumnVector(c, meta))
+              new ColumnarBatch(columnVectorArr, c.numRows())
+            case u: UnsafeRow =>
+              val joiner =
+                GenerateUnsafeRowJoiner.create(requiredSchema, Seq(meta).toStructType)
+              joiner.join(u, metadataStructUnsafeRow)
+            case i: InternalRow =>
```

Comment:
when will we hit this branch?

Reply:
AFAIK all the file sources output unsafe row. This code looks OK, if the original row is not unsafe, we don't need to create unsafe row either.